Foundations of location-sensitive polonius

lqd opened the pull request:
I'd like to land the prototype I'm describing in the [polonius project goal](https://github.com/rust-lang/rust-project-goals/issues/118). It still is incomplete and naive and terrible but it's working "well enough" to consider landing.

I'd also like to make review easier by not opening a huge PR, but have a couple small-ish ones (the +/- line change summary of this PR looks big, but >80% is moving datalog to a single place). 

This PR starts laying the foundation for that work:
- it refactors and collects 99% of the old datalog fact gen, which was spread around everywhere, into a single dedicated module. It's still present at 3 small places (one of which we should revert anyways) that are kinda deep within localized components and are not as easily extractable into the rest of fact gen, so it's fine for now.
- starts introducing the localized constraints, the building blocks of the naive way of implementing the location-sensitive analysis in-tree, which is roughly sketched out in https://smallcultfollowing.com/babysteps/blog/2023/09/22/polonius-part-1/ and https://smallcultfollowing.com/babysteps/blog/2023/09/29/polonius-part-2/ but with a different vibe than per-point environments described in these posts, just `r1@p: r2@q` constraints.
- sets up the skeleton of generating these localized constraints: converting NLL typeck constraints, and creating liveness constraints
- introduces the polonius dual to NLL MIR to help development and debugging. It doesn't do much currently but is a way to see these localized constraints: it's an NLL MIR dump + a dumb listing of the constraints, that can be dumped with `-Zdump-mir=polonius -Zpolonius=next`. Its current state is not intended to be a long-term thing, just for testing purposes -- I will replace its contents in the future with a different approach (an HTML+js file where we can more easily explore/filter/trace these constraints and loan reachability, have mermaid graphs of the usual graphviz dumps, etc).

I've started documenting the approach in this PR, I'll add more in the future. It's quite simple, and should be very clear when more constraints are introduced anyways.

r? @matthewjasper 

Best reviewed per commit so that the datalog move is less bothersome to read, but if you'd prefer we separate that into a different PR, I can do that (and michael has offered to review these more mechanical changes if it'd help).

lqd committed:
introduce localized outlives constraints

these are the basic blocks of the naive polonius constraint graph
implementation.

lqd committed:
add general documentation on the polonius module

this describes the rough algorithm using the localized constraint graph

lqd committed:
set up skeleton for localized constraints conversion

lqd committed:
extract main NLL MIR dump function

this will allow calling from polonius MIR

lqd committed:
introduce beginnings of polonius MIR dump

This is mostly for test purposes to show the localized constraints until
the MIR debugger is set up.

lqd commented:
This should all be gated away and deactivatd by default, but let's check: @bors try @rust-timer queue.

rust-timer commented:
Error occurred while parsing comment: Invalid command argument `.` (there may be no spaces around the `=` character)

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bors commented:
:hourglass: Trying commit 1331ccbfb7a0b69975381a55cd96942af737c043 with merge 0b8e535a30719ca2ebdd3c9ebb1ea3f99530bd26...
<!-- homu: {"type":"TryBuildStarted","head_sha":"1331ccbfb7a0b69975381a55cd96942af737c043","merge_sha":"0b8e535a30719ca2ebdd3c9ebb1ea3f99530bd26"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/12319335568/job/34390814766)
Build commit: 0b8e535a30719ca2ebdd3c9ebb1ea3f99530bd26 (`0b8e535a30719ca2ebdd3c9ebb1ea3f99530bd26`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/12319335568/job/34390814766"},"merge_sha":"0b8e535a30719ca2ebdd3c9ebb1ea3f99530bd26"} -->

rust-timer commented:
Queued 0b8e535a30719ca2ebdd3c9ebb1ea3f99530bd26 with parent 4847d6a9d07d4be9ba3196f6ad444af2d7bdde72, future [comparison URL](https://perf.rust-lang.org/compare.html?start=4847d6a9d07d4be9ba3196f6ad444af2d7bdde72&end=0b8e535a30719ca2ebdd3c9ebb1ea3f99530bd26).
There are currently 0 preceding artifacts in the [queue](https://perf.rust-lang.org/status.html).
It will probably take at least ~1.2 hours until the benchmark run finishes.
<!-- rust-timer: temporary -->

rust-timer commented:
Finished benchmarking commit (0b8e535a30719ca2ebdd3c9ebb1ea3f99530bd26): [comparison URL](https://perf.rust-lang.org/compare.html?start=4847d6a9d07d4be9ba3196f6ad444af2d7bdde72&end=0b8e535a30719ca2ebdd3c9ebb1ea3f99530bd26&stat=instructions:u).

### Overall result: no relevant changes - no action needed


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

@bors rollup=never
@rustbot label: -S-waiting-on-perf -perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=4847d6a9d07d4be9ba3196f6ad444af2d7bdde72&end=0b8e535a30719ca2ebdd3c9ebb1ea3f99530bd26&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=4847d6a9d07d4be9ba3196f6ad444af2d7bdde72&end=0b8e535a30719ca2ebdd3c9ebb1ea3f99530bd26&stat=max-rss)
<details>
<summary>Results (primary 4.6%, secondary 2.7%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 4.6% | [4.6%, 4.6%] | 1     |
| Regressions ❌ <br /> (secondary)  | 2.7% | [2.7%, 2.7%] | 1     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 4.6% | [4.6%, 4.6%] | 1     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=4847d6a9d07d4be9ba3196f6ad444af2d7bdde72&end=0b8e535a30719ca2ebdd3c9ebb1ea3f99530bd26&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.

### [Binary size](https://perf.rust-lang.org/compare.html?start=4847d6a9d07d4be9ba3196f6ad444af2d7bdde72&end=0b8e535a30719ca2ebdd3c9ebb1ea3f99530bd26&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 770.238s -> 768.851s (-0.18%)
**Artifact size**: 330.40 MiB -> 330.38 MiB (-0.00%)

jackh726 commented:
Started the review this, but need to step away. Thoughts:

First, do you have a branch with all the changes you've made so far? Could be good to link it if so for context.

Second, this is still a "big" PR, definitely worth splitting into a couple, I think.

That being said, reviewed the first 2 commits and r+ for me on them (if you want to split them out). I'll keep going through the commits as I get time.

lqd commented:
> First, do you have a branch with all the changes you've made so far? Could be good to link it if so for context.

Yes and no, I have a messy older incomplete one and a bunch of local changes because I'm still working on it, so I'm basically cleaning them up as I go to open PRs. 

> Second, this is still a "big" PR, definitely worth splitting into a couple, I think.

Ok. I'll split the first 2 commits out now, it should simplify this PR a bit. I can split the rest of the old polonius' fact gen later if you or michael want to look at it.

jackh726 commented on compiler/rustc_borrowck/src/polonius/legacy/mod.rs:
```diff
@@ -205,3 +210,31 @@ pub(crate) fn emit_drop_facts<'tcx>(
+    location_table: &LocationTable,
+    all_facts: &mut Option<AllFacts>,
+) {
+    if let Some(facts) = all_facts {
```
Nit: Would it be better to make this and other similar functions take a `&mut AllFacts` and move this some check out of the function (especially now that you're grouping the function calls together)?

jackh726 replied:
Actually, I see you've done this, so this is resolved

jackh726 reviewed (approved):
Have reviewed first 8 commits (through "improve consistency..."). ~A small nit, but otherwise~ r=me if you want to split them out.

jackh726 commented on compiler/rustc_borrowck/src/polonius/mod.rs:
```diff
@@ -1,3 +1,38 @@
+//!
+//! More information on this simple approach can be found in the following links, and in the future
+//! in the rustc dev guide:
+//! - <https://smallcultfollowing.com/babysteps/blog/2023/09/22/polonius-part-1/>
```
It would be good to move the details to rustc_dev_guide at some point - Niko's blog is great but should make sure there is thorough docs somewhere more durable.

lqd replied:
Agreed, I plan on writing this soon.

jackh726 replied:
I think the doc comment here is fine!

jackh726 commented on compiler/rustc_borrowck/src/nll.rs:
```diff
@@ -135,6 +139,18 @@ pub(crate) fn compute_regions<'a, 'tcx>(
+    // constraints.
+    let mut localized_outlives_constraints = LocalizedOutlivesConstraintSet::default();
+    if infcx.tcx.sess.opts.unstable_opts.polonius.is_next_enabled() {
+        polonius::create_localized_constraints(
```
Is there any reason that this function can't just return a `LocalizedOutlivesConstraintSet` instead of taking it as a mutable reference?

When polonious=next is *not* enabled, I think it makes sense to have the local var here be `None`, and Some otherwise (but this point is just a nit).

lqd replied:
No fundamental reason, so I've made this an Option.

jackh726 commented on compiler/rustc_borrowck/src/polonius/mod.rs:
```diff
@@ -35,5 +35,146 @@
+    body: &Body<'tcx>,
+    localized_outlives_constraints: &mut LocalizedOutlivesConstraintSet,
+) {
+    if !tcx.sess.opts.unstable_opts.polonius.is_next_enabled() {
```
This flag is checked twice: once in the caller, and one here. One of them should probably be removed (I have no real preference which).

jackh726 commented on compiler/rustc_borrowck/src/polonius/mod.rs:
```diff
@@ -35,5 +35,146 @@
+    for outlives_constraint in outlives_constraints {
+        match outlives_constraint.locations {
+            Locations::All(_) => {
+                // FIXME: for now, turn logical constraints holding at all points into physical
```
Very minor nit: the FIXME could be much shorter with only the action, i.e. " encode this into *traversal* instead. And the rest of it moved to a non-FIXME comment

lqd replied:
Done. I've clarified the comment and extracted the actionable part.

jackh726 commented on compiler/rustc_borrowck/src/polonius/mod.rs:
```diff
@@ -35,5 +35,146 @@
+    localized_outlives_constraints: &mut LocalizedOutlivesConstraintSet,
+) {
+    // Universal regions are semantically live at all points.
+    // FIXME: We always have universal regions but they're not always (or often) involved in the
```
Similarly minor nit as above: it's not clear what the action here is supposed to be

lqd replied:
Same as above, I've clarified the comment and extracted the actionable part.

jackh726 reviewed (commented):
Cursory review of commits 9 onward.

Overall they look fine, but didn't think *too* deeply so not a full approval.

lqd committed:
address review comments

- move constraints to an Option
- check `-Zpolonius=next` only once
- rewrite fixme comments to make the actionable part clear

jackh726 commented:
I will come back to this for approval of the last few commits once the PR with the first 8 commits lands.

lqd commented:
I rebased after #134378 landed.

jackh726 commented:
r? jackh726

@bors r+

bors commented:
:pushpin: Commit aeb3d103a6913115996d0ddb9d550016ea403488 has been approved by `jackh726`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=jackh726 aeb3d103a6913115996d0ddb9d550016ea403488 -->
<!-- homu: {"type":"Approved","sha":"aeb3d103a6913115996d0ddb9d550016ea403488","approver":"jackh726","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit aeb3d103a6913115996d0ddb9d550016ea403488 with merge 426d1734238e3c5f52e935ba4f617f3a9f43b59d...
<!-- homu: {"type":"BuildStarted","head_sha":"aeb3d103a6913115996d0ddb9d550016ea403488","merge_sha":"426d1734238e3c5f52e935ba4f617f3a9f43b59d"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/12448514500/job/34755060805)
Approved by: jackh726
Pushing 426d1734238e3c5f52e935ba4f617f3a9f43b59d to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"jackh726","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/12448514500/job/34755060805"},"merge_sha":"426d1734238e3c5f52e935ba4f617f3a9f43b59d"} -->

rust-timer commented:
Finished benchmarking commit (426d1734238e3c5f52e935ba4f617f3a9f43b59d): [comparison URL](https://perf.rust-lang.org/compare.html?start=4c40c89c26e9a6c761e2c6d44662028e33e9cb04&end=426d1734238e3c5f52e935ba4f617f3a9f43b59d&stat=instructions:u).

### Overall result: ❌ regressions - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=4c40c89c26e9a6c761e2c6d44662028e33e9cb04&end=426d1734238e3c5f52e935ba4f617f3a9f43b59d&stat=instructions:u)
This is the most reliable metric that we have; it was used to determine the overall result at the top of this comment. However, even this metric can sometimes exhibit noise.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -    | -            | 0     |
| Regressions ❌ <br /> (secondary)  | 0.2% | [0.2%, 0.2%] | 1     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | -    | -            | 0     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=4c40c89c26e9a6c761e2c6d44662028e33e9cb04&end=426d1734238e3c5f52e935ba4f617f3a9f43b59d&stat=max-rss)
<details>
<summary>Results (primary 2.2%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 2.2% | [2.2%, 2.2%] | 1     |
| Regressions ❌ <br /> (secondary)  | -    | -            | 0     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 2.2% | [2.2%, 2.2%] | 1     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=4c40c89c26e9a6c761e2c6d44662028e33e9cb04&end=426d1734238e3c5f52e935ba4f617f3a9f43b59d&stat=cycles:u)
<details>
<summary>Results (primary -4.4%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -4.4% | [-6.7%, -2.0%] | 2     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -4.4% | [-6.7%, -2.0%] | 2     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=4c40c89c26e9a6c761e2c6d44662028e33e9cb04&end=426d1734238e3c5f52e935ba4f617f3a9f43b59d&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 760.428s -> 762.028s (0.21%)
**Artifact size**: 330.62 MiB -> 330.62 MiB (0.00%)
